### PR TITLE
Guides 2.0 - Actions # Control Flow section

### PR DESCRIPTION
**Pull request by swilgosz**
Resolves: https://trello.com/c/6q08oFHM/304-actions-control-flow

Extracted from: https://github.com/hanami/guides/pull/101

### CODE REVIEW

File: content/v2.0/actions/control-flow.md [hunk 1]
```diff
@@ -0,0 +1,198 @@
...
+title: Control Flow
+order: 110
+aliases:
+  - "/actions/control-flow"
+---
+
+## Callbacks
+
+If we want to execute some logic before and/or after `#handle` is executed, we can use a callback.
+Callbacks are useful to declutter code for common tasks like checking if a user is signed in, set a record, handle 404 responses or tidy up the response.
```
Comment (solnic) [inline content/v2.0/actions/control-flow.md:11]: What do you mean by `set a record`?

File: content/v2.0/actions/control-flow.md [hunk 2]
```diff
@@ -0,0 +1,198 @@
...
+
+They are bound to the instance context of the action.
+
+```ruby
+# app/actions/books/show.rb
+module Bookshelf
+  module Actions
+    module Books
+      class Show < Action
+        before { |req, res| halt 422, req.params.errors.to_h unless req.params.valid? }
```
Comment (solnic) [inline content/v2.0/actions/control-flow.md:59]:
```suggestion
        before { |request, response| halt 422, request.params.errors.to_h unless request.params.valid? }
```

**Review by solnic — CHANGES_REQUESTED** [review]
I stopped half-way because it seems like this is quite outdated. ie it still uses `Hanami::Application`, refers to `req` and `res` etc.

Any chance you could revisit this before we can review it? 🙏🏻

File: content/v2.0/actions/control-flow.md (hunk 1 above)
Comment (swilgosz) [reply content/v2.0/actions/control-flow.md:11]: I see it like: fetching a record to use in order to process/check authorize access to. This is not changed from guides 1.3.

**Comment by swilgosz** [comment]
OMG. Thanks, @solnic , I've accidentially commited `mime-types` which is not started.

File: content/v2.0/actions/control-flow.md (hunk 1 above)
Comment (swilgosz) [reply content/v2.0/actions/control-flow.md]: I have removed this mention for clarity. It's not necessary to show the point.

**Comment by swilgosz** [comment]
PR is ready to review again.

File: content/v2.0/actions/control-flow.md [hunk 3]
```diff
@@ -0,0 +1,198 @@
...
+  - "/actions/control-flow"
+---
+
+## Callbacks
+
+If we want to execute some logic before and/or after `#handle` is executed, we can use a callback.
+Callbacks are useful to declutter code for common tasks like checking if a user is signed in, handling 404 responses or tidying up the response.
+
+The corresponding DSL methods are `before` and `after`.
+These methods each accept a symbol that is the name of the method that we want to call, or an anonymous proc.
```
Comment (solnic) [inline content/v2.0/actions/control-flow.md:14]:
```suggestion
These methods each accept a symbol that is the name of the method that we want to call, or a proc.
```

File: content/v2.0/actions/control-flow.md [hunk 4]
```diff
@@ -0,0 +1,198 @@
...
+end
+```
+
+### Back
+
+Sometimes you'll want to `redirect_to` back in your browser's history so the easy way to do it
+is the following way:
+
+```ruby
+res.redirect_to request.get_header("Referer") || fallback_url
```
Comment (solnic) [inline content/v2.0/actions/control-flow.md:197]:
```suggestion
response.redirect_to request.get_header("Referer") || fallback_url
```

**Review by solnic — CHANGES_REQUESTED** [review]

File: content/v2.0/actions/control-flow.md [hunk 5]
```diff
@@ -0,0 +1,211 @@
...
+    end
+  end
+end
+```
+
+<p><img src="/v2.0/actions/404-response.png" alt="Hanami 404 response" class="img-responsive"></p>
+
+## Redirects
+
+To redirect a request to another location, call the `#redirect_to` method on the `request` object.
```
Comment (timriley) [inline content/v2.0/actions/control-flow.md:161]:
```suggestion
To redirect a request to another location, call the `#redirect_to` method on the `response` object.
```

File: content/v2.0/actions/control-flow.md [hunk 6]
```diff
@@ -0,0 +1,211 @@
...
+
+When `redirect_to` is invoked, control flow is stopped and subsequent code in the action is not executed.
```
Comment (timriley) [inline content/v2.0/actions/control-flow.md:163]:
```suggestion
When you call `redirect_to`, control flow is stopped and subsequent code in the action is not executed.
```

**Review by timriley — APPROVED** [review]
Looks good!

**Pull request merged**